doc/source/development/rfc/rfc92_wkb_only_geometries.rst

```diff
@@ -0,0 +1,228 @@
+        // to wrongly cast to a OGRPoint/OGRLineString/etc. instance.
+    };
+
+
+A new method is added to the OGRLayer class:
+
+.. code-block:: c++
+
+    /** If bRequestWKBOnlyGeometries is true, then the driver should return, in
+     *  GetNextFeature(), geometries that are instance of OGRWKBOnlyGeometry.
+     *  Only drivers for which TestCapability(OLCReadWKBGeometries) is true
+     *  are capable of this. Other drivers will error out.
+     */
+    virtual OGRErr RequestWKBOnlyGeometries(bool bRequestWKBOnlyGeometries);
+
+Two new capabilities are added at the OGRLayer level:
+
+* OLCReadWKBGeometries: a layer must return TRUE for it when the layer can
+  honour RequestWKBOnlyGeometries(true)
+* OLCWriteWKBGeometries: a layer must return TRUE for it if its CreateFeature()
+  and SetFeature() implementations support being passed OGRWKBOnlyGeometry
+  instances.
+
+ogr2ogr is modified to call RequestWKBOnlyGeometries(true) on the source layer:
+
+* if the source layer advertises OLCReadWKBGeometries
+* if the target layer advertises OLCWriteWKBGeometries
+* if no command line switch requires a "materialized" geometry.
+
+So basically this is for requests like
+"ogr2ogr out.gpkg in.gpkg [layer or SQL request] [attribute filter] [spatial filter]"
+
+C API
+-----
+
+OGR_L_RequestWKBOnlyGeometries() is added.
+
+Backward compatibility
+----------------------
+
+No issue. Only API additions
+
+Benchmarking
+------------
+
+bench_ogr_c_api
+~~~~~~~~~~~~~~~
+
+The `bench_ogr_c_api <https://github.com/OSGeo/gdal/blob/master/perftests/bench_ogr_c_api.cpp>`_
+benchmark utility which uses the C API to iterate over features and get their WKB
+representation is enhanced with a `-wkb_only_geometry` switch to call
+OGR_L_RequestWKBOnlyGeometries().
+
+On a 1.6 GB GeoPackage (nz-building-outlines.gpkg) made of 3.2 million features
+with simple polygons (typically quadrilaterals, building footprints) and 13
+attributes:
+
+- bench_ogr_c_api runs in 6.4 s
+- bench_ogr_c_api -wkb_only_geometry runs in 5.0 s
+
+==> 22% faster
+
+Other synthetic benchmarks show that the maximum speed-up is about 30% on a
+dataset with 10 millions polygonal features of 10 points each.
+
+Conversely, the gain is much more modest, or close to null, with just a few
+thousands of features that hold larger geometries (several thousands of points
+each).
+
+The gain is more in saving instantiation of OGRPolygon and OGRLinearRing
+objects that in the size of their coordinate set.
+
+ogr2ogr
+~~~~~~~
+
+::
+
+    ogr2ogr /vsimem/out.gpkg nz-building-outlines.gpkg -lco spatial_index=no
+
+
+runs in 15.8 second in WKBOnlyGeometry mode vs 19.1 second without it (master),
+hence a 17% speed-up.
+
+With spatial index creation enabled (multi-threaded), the wall clock time
+difference is within measurement noise. And for singe threaded creation, the
+WKBOnlyGeometry mode is 5% faster.
+
+Discussion
+----------
+
+Is it a good idea... ?
+
```

Comment:
I believe so. I've extensively profiled qgis' ogr provider feature reading in the past and the geometry parsing/conversion costs are a large percentage of the reading time which cannot be further optimised using current API.

For some datasets this is likely to result in a noticable rendering time improvement in desktop QGIS, and a definitely worthwhile improvement for QGIS server.